api/src/main/java/org/openmrs/Cohort.java

```diff
@@ -141,44 +145,72 @@ public Cohort(String commaSeparatedIds) {
 	 */
 	public String getCommaSeparatedPatientIds() {
 		StringBuilder sb = new StringBuilder();
-		for (Iterator<Integer> i = getMemberIds().iterator(); i.hasNext();) {
-			sb.append(i.next());
-			if (i.hasNext()) {
-				sb.append(",");
-			}
+        for (CohortMembership member : getMembers()) {
+			sb.append(member.getPatient().getPatientId());
+			sb.append(",");
 		}
 		return sb.toString();
 	}
-	
+
+	// TODO: Compare patient objects instead of IDs after removal of deprecated methods
 	public boolean contains(Patient patient) {
-		return getMemberIds() != null && getMemberIds().contains(patient.getPatientId());
+		return getMembers() != null && getMembers().stream()
+				.anyMatch(m -> m.getPatient().getPatientId().equals(patient.getPatientId()));
 	}
 	
 	public boolean contains(Integer patientId) {
-		return getMemberIds() != null && getMemberIds().contains(patientId);
+		return getMembers() != null && getMembers().stream()
+				.anyMatch(m -> m.getPatient().getPatientId().equals(patientId));
 	}
-	
+
+	public boolean contains(CohortMembership cohortMembership) {
+		return getMembers() != null && getMembers().stream()
+				.anyMatch(m -> m.equals(cohortMembership));
+	}
+
 	public String toString() {
 		StringBuilder sb = new StringBuilder("Cohort id=" + getCohortId());
 		if (getName() != null) {
-			sb.append(" name=" + getName());
+			sb.append(" name=").append(getName());
 		}
-		if (getMemberIds() != null) {
-			sb.append(" size=" + getMemberIds().size());
+		if (getMembers() != null) {
+			sb.append(" size=").append(getMembers().size());
 		}
 		return sb.toString();
 	}
-	
+
 	public void addMember(Integer memberId) {
-		getMemberIds().add(memberId);
+		this.addMembership(new CohortMembership(new Patient(memberId)));
 	}
-	
-	public void removeMember(Integer memberId) {
-		getMemberIds().remove(memberId);
+
+	public void removeMember(Patient patient) {
+		List<CohortMembership> memberToRemoveList = getMembers().stream()
+				.filter(m -> m.getPatient() == patient).collect(Collectors.toList());
+		memberToRemoveList.forEach(this::removeMembership);
 	}
-	
+
+	public void addMembership(CohortMembership cohortMembership) {
+		getMembers().add(cohortMembership);
+	}
+
+	public void removeMembership(CohortMembership cohortMembership) {
+        List<CohortMembership> memberToRemoveList = getMembers().stream()
+				.filter(m -> m.equals(cohortMembership))
+				.collect(Collectors.toList());
+		memberToRemoveList.forEach(m -> m.setEndDate(new Date()));
```

Comment:
I think this should also be able to lookup the membership with the matching patient and remove it so that a call like the one below  works too:

`Cohort.removeMembership(new CohortMembership(Context.getPatientService().getPatient(patientId)))`